experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts

```diff
@@ -156,106 +277,166 @@ export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
-			const rebasedChildChange = rebaseChild(change.childChange, overChildChange);
-			if (rebasedChildChange !== undefined) {
-				rebasedChange.childChange = rebasedChildChange;
-			} else {
-				delete rebasedChange.childChange;
-			}
-
-			return rebasedChange;
-		}
-
-		if (change.childChange !== undefined) {
+			// If we're rebasing over a fieldChange, track ChangeAtomId_s for cases where a previously existing
+			// node was restored. In that case, when we construct our child id changes, they may apply to 'self' rather than
+			// the pre-existing childId.
+			let restoredRollbackChangeId: ChangeAtomId | undefined;
+			let restoredUndoChangeId: ChangeAtomId | undefined;
 			if (over.fieldChange !== undefined) {
 				const overIntention = getIntention(
 					over.fieldChange.revision ?? overTagged.revision,
 					revisionMetadata,
 				);
-				if (change.deletedBy === undefined) {
-					// `change.childChange` refers to the node being deleted by `over`.
-					return {
-						childChange: rebaseChild(
-							change.childChange,
-							over.deletedBy === undefined ? undefined : over.childChange,
-							NodeExistenceState.Dead,
-						),
-						deletedBy: {
-							revision: overIntention,
-							localId: over.fieldChange.id,
-						},
-					};
-				} else if (over.fieldChange.newContent !== undefined) {
+
+				if (over.fieldChange.newContent !== undefined) {
 					const overContent = over.fieldChange.newContent;
-					const rebasingOverRollback =
-						overIntention === change.deletedBy.revision &&
-						over.fieldChange.id === change.deletedBy.localId;
-					const rebasingOverUndo =
-						"revert" in overContent &&
-						overContent.changeId.revision === change.deletedBy.revision &&
-						overContent.changeId.localId === change.deletedBy.localId;
-					if (rebasingOverRollback || rebasingOverUndo) {
-						// Over is reviving the node that change.childChange is referring to.
-						// Rebase change.childChange and remove deletedBy
-						// because we revived the node that childChange refers to
-						return {
-							childChange: rebaseChild(
-								change.childChange,
-								overContent.changes,
-								NodeExistenceState.Alive,
-							),
-						};
+					restoredRollbackChangeId = {
+						revision: overIntention,
+						localId: over.fieldChange.id,
+					};
+					if ("revert" in overContent) {
+						restoredUndoChangeId = overContent.changeId;
 					}
 				}
 			}
-		}
-
-		{
-			const rebasedChange = { ...change };
 
-			let overChildChange: NodeChangeset | undefined;
-			if (change.deletedBy === undefined && over.deletedBy === undefined) {
-				overChildChange = over.childChange;
+			for (const [id, childChange] of change.childChanges) {
+				if (id === "self") {
+					const overChildChange = overChildChanges.get(id);
+					if (over.fieldChange !== undefined && change.fieldChange !== undefined) {
+						// `overChange` and `childChange` refer to conceptually different nodes: each replaced the field.
```

Comment:
I'm really surprised by that comment. I'll try to step through a test to see if it helps me, but I thought I'd call it out in case you think it's wrong too.

Reply:
Yeah, that still seems wrong.
Looks like there is no existing test for this. Here's one:
```
			it("can rebase child change (field change ↷ field change)", () => {
				const baseChange: OptionalChangeset = {
					fieldChange: {
						id: brand(0),
						wasEmpty: false,
					},
					childChanges: [["self", nodeChange1]],
				};
				const changeToRebase: OptionalChangeset = {
					fieldChange: {
						id: brand(1),
						wasEmpty: false,
						newContent: { set: { type, value: "X" } },
					},
					childChanges: [["self", nodeChange2]],
				};

				const childRebaser = (
					change: NodeChangeset | undefined,
					base: NodeChangeset | undefined,
				): NodeChangeset | undefined => {
					assert.deepEqual(change, nodeChange2);
					assert.deepEqual(base, nodeChange1);
					return arbitraryChildChange;
				};

				const expected: OptionalChangeset = {
					fieldChange: {
						id: brand(1),
						wasEmpty: true,
						newContent: { set: { type, value: "X" } },
					},
					childChanges: [[{ localId: brand(0) }, arbitraryChildChange]],
				};

				const actual = optionalChangeRebaser.rebase(
					changeToRebase,
					makeAnonChange(baseChange),
					childRebaser,
					failIdAllocator,
					failCrossFieldManager,
					defaultRevisionMetadataFromChanges([]),
				);
				assert.deepEqual(actual, expected);
			});
```